neps/nep-0413.md

```diff
@@ -0,0 +1,138 @@
+Sending the signed message in a query string to an arbitrary URL (even within the correct domain) is not secure as the data can be leaked (e.g. through headers, etc). Using URL fragments instead will improve security, since [URL fragments are not included in the `Referer`](https://greenbytes.de/tech/webdav/rfc2616.html#header.referer).
+
+### NEAR Signatures
+NEAR transaction signatures are not plain Ed25519 signatures but Ed25519 signatures of a SHA-256 hash (see [near/nearcore#2835](https://github.com/near/nearcore/issues/2835)). Any protocol that signs anything with NEAR account keys should use the same signature format.
+
+## Specification
+Wallets must implement a `signMessage` method, which takes a `message` destined to a specific `receiver` and transform it into a verifiable signature.
+
+### Input Interface
+`signMessage` must implement the following input interface:
+
+```jsx
+interface SignMessageParams {
+  message: string ; // The message that wants to be transmitted.
+  receiver: string; // The receiver to whom the message is destined (e.g. "alice.near" or "myapp.com").
+  nonce: [u8; 32] ; // A nonce that uniquely identifies this instance of the message, denoted as a 32 bytes array (a fixed `Buffer` in JS/TS).
+  callbackUrl?: string; // Optional, applicable to browser wallets (e.g. MyNearWallet). The URL to call after the signing process. Defaults to `window.location.href`.
+}
+```
+
+### Structure
+`signMessage` must embed the input `message`, `receiver` and `nonce` into the following predefined structure:
+
+```rust
+struct Payload {
+  message: string; // The same message passed in `SignMessageParams.message`
+  receiver: string; // The same receiver passed in `SignMessageParams.receiver`
+  nonce: [u8; 32]; // The same nonce passed in `SignMessageParams.nonce`
+}
+```
+
+### Signature
+In order to create a signature, `signMessage` must:
+1. Convert the `Payload` into its [JSON JCS](https://www.rfc-editor.org/rfc/rfc8785) string representation (i.e. a JSON string, with alphabetically ordered attributes).
+2. Prepend the `NEP0413:` string to the result from step 1.
+3. Compute the `SHA256` hash of the result from step 2.
+4. Sign the resulting `SHA256` hash from step 3 using a **full-access** key.
+
+> If the wallet does not hold any `full-access` keys, then it must return an error.
+
+### Example
+Assuming that the `signMessage` method was invoked, and that:
+- The input `message` is `"hi"`
+- The input `receiver` is `"myapp.com"`
+- The input `nonce` is `[0,...,31]`
+- The wallet stores a full-access private key
+
+The wallet must construct and sign the following `SHA256` hash:
+
+```jsx
+sha256.hash(`NEP0413:` + `{"message":"hi","receiver":"myapp.com","nonce":"[0,...,31]"}`)
+```
+
+### Output Interface
+`signMessage` must return an object containing the **base64** representation of the `signature`, and all the data necessary to verify such signature. 
+
+```jsx
+interface SignedMessage {
```

Comment:
Should `SignedMessage` also include the un-signed string that was hashed and then signed, so that the caller can easily verify the signature is correct without needing to know the details of how to serialize the `SignMessageParams` into the `Payload` struct as JSON with the added prefix of `NEP0413:`?

Reply:
We want to return the minimum and necessary to validate the signature. While it could be convenient to return the signed string, it is not strictly necessary since anyway the logic is well defined. Therefore it is better to not include it.

This is actually the case of Ethereum, where they also sign a `head:message`, and they return only the signature, without returning the actually signed string.

Reply:
I'd suggest keeping in mind the separation of concerns principle. The `signMessage` function requires the `data` returns the `signature`. On the reverse `verify` requires the `signature` and `data` and returns the verification result. A separate function for example `createSignatureData` create the hash/message that will be signed/verified by the underlying private/public key. The `sign` and `verify` functions would call `createSignatureData`. Also developers can use `createSignatureData` if they wish to. I think this would represent a good balance between ease of use, access to the signing message and having `sign` return only what is needed.

Reply:
Particular implementation can take this route if they decide, as the reference says what the input and output should be, not how the exact code must look like.

Reply:
I agree with @gagdiez, my previous comment should be taken as a suggestion rather than part of the spec. I think it helps too imagine the implementation to show why returning the minimum is acceptable but should be anything more than that.

Reply:
@gagdiez in the examples the [verifySignature](https://github.com/gagdiez/near-login/blob/main/authenticate/wallet-authenticate.js#L19) function requires the un-signed `message` to reconstruct the payload.

The value of the `message` in case of Browser Wallets that require redirect will be lost and when the user gets redirected back to the dApp.

How can we reconstruct the payload as in the above function in this case, or how can we verify the signature?

Reply:
@kujtimprenkuSQA in this case, you will need to ask the user to pass the `message` that they signed.